# 4.6 Merging and exporting data exercise

#### Import libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

#### Create path

In [9]:
path = r'C:\Users\arpit\Documents\CareerFoundry\20250505 Instacart Basket Analysis'

#### Import dataframes

In [12]:
# Import dataframe

orders_products_combined = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'orders_products_combined.pkl'))

df_prods = pd.read_csv(os.path.join(path, '02 data' , 'prepared data' , 'products_cleaned.csv'), index_col = 0)

In [14]:
# Check the shape of the imported dataframe

orders_products_combined.shape

(32434489, 11)

#### Understanding the basic information of the dataframes

In [18]:
# Check dataframe structure

orders_products_combined.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [20]:
# Check the shape

df_prods.shape

(49672, 5)

In [22]:
# Check dataframe structure

df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

#### Merging dataframe

**Evaluation**
There's a key column 'product_id' observed among these two dataframes, indicating a merge is possible, however, as for which type of the merge should be applied to depends on the information we expect to see in the final dataframe. Since the information wouldn't be much useful if it doesn't exist in both dataframes, therefore, the 'inner' merge will be a suitable opion.

In [27]:
# test merge without overwriting with merge flag 'match'

pd.merge(orders_products_combined , df_prods , on = ['product_id'] , how = 'inner', indicator = 'match' )

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                  0   2539329        1             1                   2   
1                  0   2539329        1             1                   2   
2                  0   2539329        1             1                   2   
3                  0   2539329        1             1                   2   
4                  0   2539329        1             1                   2   
...              ...       ...      ...           ...                 ...   
32404854     3421081   2977660   206209            13                   1   
32404855     3421081   2977660   206209            13                   1   
32404856     3421081   2977660   206209            13                   1   
32404857     3421081   2977660   206209            13                   1   
32404858     3421081   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  \
0                         1          0   both   
1                         2          0   both   
2                         3          0   both   
3                         4          0   both   
4                         5          0   both   
...                     ...        ...    ...   
32404854                  5          1   both   
32404855                  6          0   both   
32404856                  7          0   both   
32404857                  8          0   both   
32404858                  9          0   both   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32404854                             Tomato Paste         9              9   
32404855          Brownie Crunch High Protein Bar         3             19   
32404856    High Protein Bar Chunky Peanut Butter         3             19   
32404857      Chocolate Peanut Butter Protein Bar         3             19   
32404858      Roasted & Salted Shelled Pistachios       117             19   

          prices match  
0            9.0  both  
1           12.5  both  
2            4.4  both  
3            4.7  both  
4            1.0  both  
...          ...   ...  
32404854     5.6  both  
32404855     5.9  both  
32404856     4.2  both  
32404857     4.9  both  
32404858     1.7  both  

[32404859 rows x 16 columns]

**Evaluation**
As the previous exercise of merged dataframe has created 'merge' column as the flag which would be no use for the new merge of 'orders_products_combined' and 'df_prods', for better presetation, the 'merge' column from the previous merge should be dropped after the new merge has been carried out.

In [46]:
# actual merge of 'orders_products_combined' and '_df_pords'

df_ords_prods_merge = orders_products_combined.merge(df_prods , on = ['product_id'] , indicator = 'match' )

In [47]:
# drop the flag column '_merge' from the previous merge

df_ords_prods_merge.drop('_merge', axis = 1 )

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                  0   2539329        1             1                   2   
1                  0   2539329        1             1                   2   
2                  0   2539329        1             1                   2   
3                  0   2539329        1             1                   2   
4                  0   2539329        1             1                   2   
...              ...       ...      ...           ...                 ...   
32404854     3421081   2977660   206209            13                   1   
32404855     3421081   2977660   206209            13                   1   
32404856     3421081   2977660   206209            13                   1   
32404857     3421081   2977660   206209            13                   1   
32404858     3421081   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered  \
0                         1          0   
1                         2          0   
2                         3          0   
3                         4          0   
4                         5          0   
...                     ...        ...   
32404854                  5          1   
32404855                  6          0   
32404856                  7          0   
32404857                  8          0   
32404858                  9          0   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32404854                             Tomato Paste         9              9   
32404855          Brownie Crunch High Protein Bar         3             19   
32404856    High Protein Bar Chunky Peanut Butter         3             19   
32404857      Chocolate Peanut Butter Protein Bar         3             19   
32404858      Roasted & Salted Shelled Pistachios       117             19   

          prices match  
0            9.0  both  
1           12.5  both  
2            4.4  both  
3            4.7  both  
4            1.0  both  
...          ...   ...  
32404854     5.6  both  
32404855     5.9  both  
32404856     4.2  both  
32404857     4.9  both  
32404858     1.7  both  

[32404859 rows x 15 columns]

In [50]:
# Check output
df_ords_prods_merge.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered _merge                             product_name  aisle_id  \
0          0   both                                     Soda        77   
1          0   both  Organic Unsweetened Vanilla Almond Milk        91   
2          0   both                      Original Beef Jerky        23   
3          0   both               Aged White Cheddar Popcorn        23   
4          0   both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match  
0              7     9.0  both  
1             16    12.5  both  
2             19     4.4  both  
3             19     4.7  both  
4             17     1.0  both

In [54]:
df_ords_prods_merge.shape

(32404859, 16)

#### Confirm the results of the merge using the merge flag

In [58]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

After merging the resulting dataframe has 32,404,859 rows, and each of those rows have information found in both input data sets, as we used an inner join for the purposes of this project

#### Export the newly created dataframe

In [62]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))